src/System.Runtime.Extensions/src/System/Net/WebUtility.cs

```diff
@@ -367,9 +370,37 @@ public static string UrlEncode(string value)
             if (string.IsNullOrEmpty(value))
                 return value;
 
-            byte[] bytes = Encoding.UTF8.GetBytes(value);
-            byte[] encodedBytes = UrlEncode(bytes, 0, bytes.Length, false /* alwaysCreateNewReturnValue */);
-            return Encoding.UTF8.GetString(encodedBytes, 0, encodedBytes.Length);
+            int cSafe = 0;
+            int cSpaces = 0;
+            for (int i = 0; i < value.Length; i++)
+            {
+                char ch = value[i];
+                if (ch == ' ')
+                {
+                    cSpaces++;
+                }
+                else if (IsUrlSafeChar(ch))
+                {
+                    cSafe++;
+                }
+            }
+
+            if (cSpaces == 0 && cSafe == value.Length)
```

Comment:
Are both of these conditions necessary?  If cSafe == value.Length, how could cSpaces not be 0?